### DISCUSSION TITLE
Input Field Validations

### DISCUSSION THREAD

**Discussion by we-wake-act21** [keycloak/keycloak#32334] [Q&A]
Hi,

While using Keycloak, I've noticed that none of the input fields have input validation, 
allowing special characters and improper formats to be accepted. 

How can I configure Keycloak to restrict these improper input formats and patterns?

![image](https://github.com/user-attachments/assets/5d2576d5-a6fa-4103-8797-2f2ba38e58f7)

**Comment by danielFesenmeyer** [comment]
Afaik, currently you can only add validation for user profile fields, not for other entities like roles.

Apart from that, Keycloak applies [output encoding ](https://cheatsheetseries.owasp.org/cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.html#output-encoding) for all data, except for special "html fields" (which are described accordingly in the UI).
This is a good measure against XSS attacks. 

If you want to have even more protection by means of [input validation](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html), you might consider using a Web Application Firewall.
